lib/pharos/transport/ssh.rb

```diff
@@ -35,12 +43,15 @@ def connect(**options)
           begin
             @session = session_factory.start(@host, @user, @opts.merge(options).merge(non_interactive: non_interactive))
             logger.debug "Connected"
+            class_mutex.unlock if class_mutex.locked? && class_mutex.owned?
```

Comment:
There is already `synchronize` (at line 40)...?

Reply:
It's instance level.

An instance should not try to connect at the same time on multiple threads. Multiple instances can each connect at the same time because they don't share a state.

The class level mutex protects against any of the instances trying to prompt at the same time.

It's not using `class_mutex.synchronize` because you can't  call retry from a block (retry never returns).

Reply:
Ok, makes sense.